x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticMatchQueryRewriteInterceptor.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.queries;
+
+import org.elasticsearch.action.ResolvedIndices;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.InferenceFieldMetadata;
+import org.elasticsearch.features.NodeFeature;
+import org.elasticsearch.index.mapper.IndexFieldMapper;
+import org.elasticsearch.index.query.AbstractQueryBuilderWrapper;
+import org.elasticsearch.index.query.BoolQueryBuilder;
+import org.elasticsearch.index.query.MatchQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryRewriteContext;
+import org.elasticsearch.index.query.QueryRewriteInterceptor;
+import org.elasticsearch.index.query.TermQueryBuilder;
+import org.elasticsearch.index.query.TermsQueryBuilder;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+public class SemanticMatchQueryRewriteInterceptor implements QueryRewriteInterceptor {
+
+    public static final NodeFeature SEMANTIC_MATCH_QUERY_REWRITE_INTERCEPTION_SUPPORTED = new NodeFeature(
+        "search.semantic_match_query_rewrite_interception_supported"
+    );
+
+    public SemanticMatchQueryRewriteInterceptor() {}
+
+    @Override
+    public QueryBuilder rewrite(QueryRewriteContext context, QueryBuilder queryBuilder) {
+
+        if (queryBuilder instanceof MatchQueryBuilder == false) {
```

Comment:
Why is this needed since we registered that the interceptor is for a single query. Maybe we should enforce that and add the template on the  `QueryRewriteInterceptor<T extends QueryBuilder>` and then rewrite becomes `rewrite(QueryRewriteContext context, T queryBuilder)`?

Reply:
You're right, missed this cleanup so far, I will make sure to address.

Reply:
I did some time boxed refactoring on this, and changing the template requires a lot of other downstream changes, not all of which I like. I'm leaning toward keeping this for now, we could always optimize it in the future.

Reply:
I am fine if we don't do the templating but even then I don't think this `instanceof` is needed. Rewrite is called only if the query name matches so it must be an instance of `MatchQueryBuilder`? Maybe an assert is enough to ensure that we don't break this behaviour in the future?

Reply:
I'm good changing to assert. Updated: 7383ac6a07879810fc597e0de7809d613effdb17